### PR TITLE
sig-scheduling: Scheduling queue in kube-scheduler

### LINKED ISSUE #5485: devel/sig-scheduling: Document the scheduling queues and events
refer #5237 

Document the different scheduling queues (active, backoff, unschedulable) and how pods move from one to another.

/sig scheduling

### PR DESCRIPTION
**Pull request by ingvagabund**
**Which issue(s) this PR fixes**:

Fixes #5485

Based on https://docs.google.com/document/d/1gy1b6dceiKCdM3oFalfJNfmiHtb46bTUz0a2pC_W5Bo/edit#

### CODE REVIEW

**Comment by ingvagabund** [comment]
cc @adtac @damemi @alculquicondor

**Comment by alculquicondor** [comment]
We should start thinking of linking to some of the docs from scheduler.md

But fine to do in a separate PR, as that file is highly outdated.

/lgtm

File: contributors/devel/sig-scheduling/scheduler_queues.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+## Active queue (heap)
+
+A queue with the highest priority pod at the top by default. The ordering
+can be customized via QueueSort extension point. Newly created pods, with empty `.spec.nodeName`,
+are added to the queue as they come. In each scheduling cycle the scheduler takes
+one pod from the queue and tries to schedule it. In case the scheduling algorithm
+fails (e.g. plugins error, binding error), the pod is moved to the unschedulable queue.
+Or, moved to the backoff queue if a move request was issued at the same or newer time.
+The move request signals a move of pods from unschedulable to active, resp. backoff queue.
```
Comment (ahg-g) [inline contributors/devel/sig-scheduling/scheduler_queues.md:40]: what is "resp."?

**Comment by damemi** [comment]
This is a nit, and only affects people who choose to have a dark background, but would it be easy for you to give the graph a white background? This is what it looks like on dark theme:
Screen Shot 2021-02-25 at 12 07 48 PM

File: contributors/devel/sig-scheduling/scheduler_queues.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+so the pods stay in the backoff queue long enough to avoid flooding the active queue
+with pods failing too often to be scheduled.
+
+## Unschedulable queue (map)
+Queue keeping all pods that failed to be scheduled and were not subject to a move request.
+Pods are kept in the queue until a move request is issued.
+
+## Moving request
+
+Moving request triggers an operation responsible for moving pods from
```
Comment (ahg-g) [inline contributors/devel/sig-scheduling/scheduler_queues.md:65]: lets use event rather than "operation" when applicable.

**Review by ahg-g — COMMENTED** [review]
Thanks Jan,

The graph looks good, but one thing we can do is remove what you call `cond`, simply indicate that the pod will move to unschedulableQ on failure. That cond is a minor optimization that doesn't help in the understanding of how pods move between queues.

File: contributors/devel/sig-scheduling/scheduler_queues.md (hunk 1 above)
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduler_queues.md:40]: respectively

File: contributors/devel/sig-scheduling/scheduler_queues.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+such as existence of a persistent volume, compliance with pod anti-affinity rules
+or toleration of node taints, the mechanism needs to be able to postpone
+the scheduling action until the cluster may meet all the conditions for
+the successful scheduling. The mechanism relies on three queues:
+- active ([activeQ](https://github.com/kubernetes/kubernetes/blob/4cc1127e9251fff364d5c77e2a9a9c3ad42383ab/pkg/scheduler/internal/queue/scheduling_queue.go#L130)): providing pods for immediate scheduling
+- unschedulable ([unschedulableQ](https://github.com/kubernetes/kubernetes/blob/4cc1127e9251fff364d5c77e2a9a9c3ad42383ab/pkg/scheduler/internal/queue/scheduling_queue.go#L135)): for parking pods which are waiting for certain condition(s) to happen
+- backoff ([podBackoffQ](https://github.com/kubernetes/kubernetes/blob/4cc1127e9251fff364d5c77e2a9a9c3ad42383ab/pkg/scheduler/internal/queue/scheduling_queue.go#L133)): exponentially postponing pods which failed
+  to be scheduled (e.g. volume still getting created) but are expected to get scheduled eventually.
+
+In addition, the scheduling queue mechanism has two synchronous flushing goroutines
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply contributors/devel/sig-scheduling/scheduler_queues.md:15]: I used word periodical instead.

**Comment by ingvagabund** [comment]
@damemi @ahg-g updated the image as recommended. Anything else to address? If not, I will squash the commits.

**Comment by ahg-g** [comment]
thanks, please squash.

**Pull request merged**